### PR TITLE
[SDK] Partial Animations API

### PR DESCRIPTION
**Pull request by stevenvergenz**
First usable iteration of the new animations API. Tandem PR with microsoft/mixed-reality-extension-unity#121.

* Add design doc to /design.
* Rewrite gltf-animation-test as animation-test.
* Add `Animation` type, which is the new repository of all animation state and commands.
* Make `context.conn.quality.latencyMs` use the values from the authoritative peer. Used for `animation.play()`
* Use high-precision timer when calculating network latency.
* Add current server time to heartbeat for anim syncing.
* Add animation definitions to `object-spawned` messages.
* Add client version to client class.

### CODE REVIEW

File: packages/sdk/src/adapters/multipeer/protocols/clientSync.ts [hunk 1]
```diff
@@ -202,15 +202,31 @@ export class ClientSync extends Protocols.Protocol {
...
-	public 'stage:sync-animations' = async () => {
+	public 'stage:sync-animations' = () => {
+		// sync new-style animations
+		for (const anim of this.client.session.animations) {
+			if (anim.update) {
+				super.sendMessage(anim.update);
+			}
+		}
+
+		// sync legacy animations
```
[the remark this answers is not in the document]
Comment (stevenvergenz) [reply packages/sdk/src/adapters/multipeer/protocols/clientSync.ts]: I think the sync-animations stage is also needed by interpolations.

---
In reply to: [363516809](https://github.com/microsoft/mixed-reality-extension-sdk/pull/475#discussion_r363516809) [](ancestors = 363516809)

File: packages/sdk/src/animation/animation.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+		this.weight = 0;
+	}
+
+	private _finished: ExportedPromise = null;
+	/** @returns A promise that resolves when the animation completes. This only occurs if the wrap mode is set
+	 * to "Once". The promise is not resolved if the animation is stopped manually.
+	 */
+	public finished(): Promise<void> {
+		if (this._finished) {
+			return this._finished.original;
```
[the remark this answers is not in the document]
Comment (stevenvergenz) [reply packages/sdk/src/animation/animation.ts]: `this._finished` is not a resolved promise, it is an in-progress promise. So not equivalent to `Promise.resolve()`.

---
In reply to: [363522699](https://github.com/microsoft/mixed-reality-extension-sdk/pull/475#discussion_r363522699) [](ancestors = 363522699)

File: packages/sdk/src/adapters/multipeer/rules.ts [hunk 3]
```diff
@@ -4,6 +4,7 @@
  */
 
 import deepmerge from 'deepmerge';
+import semver from 'semver';
```
Comment (tombuMS) [inline packages/sdk/src/adapters/multipeer/rules.ts]: I don't see semver being used any longer here after the backwards compat removal.  Safe to remove this?

File: packages/sdk/src/adapters/multipeer/session.ts [hunk 4]
```diff
@@ -170,6 +180,11 @@ export class Session extends EventEmitter {
 			client.setAuthoritative(false);
 		}
 
+		// forward connection quality metrics
+		if (this.conn instanceof EventedConnection) {
```
Comment (eanders-MS) [inline packages/sdk/src/adapters/multipeer/session.ts]: I don't think this is going to work. The connection quality object cannot be shared. Both connections will update the connection quality, making the values meaningless. What are you trying to accomplish? Maybe there's another way.

File: packages/sdk/src/adapters/multipeer/protocols/clientSync.ts [hunk 5]
```diff
@@ -202,15 +202,31 @@ export class ClientSync extends Protocols.Protocol {
...
 		const authoritativeClient = this.client.session.authoritativeClient;
 		if (!authoritativeClient) {
```
Comment (eanders-MS) [inline packages/sdk/src/adapters/multipeer/protocols/clientSync.ts]: All code in this method after your additions above should be removed. This is the source of the "timed out waiting for sync-animations" bug.

File: packages/sdk/src/adapters/multipeer/rules.ts [hunk 6]
```diff
@@ -235,6 +236,7 @@ const CreateActorRule: Rule = {
 			message: Message<Payloads.CreateActorCommon>
 		) => {
 			session.cacheInitializeActorMessage(message);
+			session.cacheAnimationCreationRequest(message);
```
Comment (eanders-MS) [inline packages/sdk/src/adapters/multipeer/rules.ts]: How is this actor creation message also an animation creation message?

File: packages/sdk/src/animation/animation.ts [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+import { AnimationWrapMode, InternalAnimation } from '.';
+import { Patchable } from '../types/patchable';
+import { Context } from '../types/runtime';
+import { ExportedPromise } from '../utils/exportedPromise';
+import readPath from '../utils/readPath';
+
+/** A serialized animation definition */
+export interface AnimationLike {
+	/** Generated unique ID of this animation */
+	id: Guid;
```
[the remark this answers is not in the document]
Comment (stevenvergenz) [reply packages/sdk/src/animation/animation.ts]: I agree! Just wanted to test it out during my development to see if it wasn't too cumbersome. Once this is merged I'll convert everything else in a separate PR.

---
In reply to: [364918638](https://github.com/microsoft/mixed-reality-extension-sdk/pull/475#discussion_r364918638) [](ancestors = 364918638)

File: packages/sdk/src/animation/animation.ts [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+	 */
+	public play(reset = false) {
+		// no-op if already playing
+		if (this.isPlaying) { return; }
+
+		// Getter for basis time converts the internal _time var into the corresponding basis time,
+		// so reassigning it writes this converted time back into the internal _basisTime var.
+		this.basisTime = (reset ? Date.now() : this.basisTime)
+			// start slightly in the future so we don't always skip over part of the animation.
+			+ this.context.conn.quality.latencyMs.value / 1000;
```
[the remark this answers is not in the document]
Comment (stevenvergenz) [reply packages/sdk/src/animation/animation.ts]: It is a pipe, but there is code elsewhere to duplicate the authoritative peer's latency metrics into the pipe.

---
In reply to: [364927044](https://github.com/microsoft/mixed-reality-extension-sdk/pull/475#discussion_r364927044) [](ancestors = 364927044)

File: packages/sdk/src/adapters/multipeer/protocols/clientSync.ts (hunk 5 above)
Comment (stevenvergenz) [reply packages/sdk/src/adapters/multipeer/protocols/clientSync.ts]: Agreed, but I can't until I've replaced interpolations. They rely on this code for sync still. All other animations use the new stuff though.

---
In reply to: [364937498](https://github.com/microsoft/mixed-reality-extension-sdk/pull/475#discussion_r364937498) [](ancestors = 364937498)

File: packages/sdk/src/adapters/multipeer/session.ts (hunk 4 above)
Comment (stevenvergenz) [reply packages/sdk/src/adapters/multipeer/session.ts]: This is forwarding the latency updates from the authoritative peer to the pipe. See the changes to EventedConnection and Pipe.

---
In reply to: [364934306](https://github.com/microsoft/mixed-reality-extension-sdk/pull/475#discussion_r364934306) [](ancestors = 364934306)

File: packages/sdk/src/adapters/multipeer/rules.ts (hunk 6 above)
Comment (stevenvergenz) [reply packages/sdk/src/adapters/multipeer/rules.ts]: Animations can be created from either CreateAnimation or CreateFromPrefab messages.

---
In reply to: [364938287](https://github.com/microsoft/mixed-reality-extension-sdk/pull/475#discussion_r364938287) [](ancestors = 364938287)

File: packages/sdk/src/adapters/multipeer/session.ts (hunk 4 above)
Comment (eanders-MS) [reply packages/sdk/src/adapters/multipeer/session.ts]: When the connection quality tracker object shared, both connections will record latency samples, trashing the calculated average.

**Pull request merged**